src/pubsub.c

```diff
@@ -95,6 +95,28 @@ pubsubtype pubSubShardType = {
     .messageBulk = &shared.smessagebulk,
 };
 
+uint64_t clientHash(const void *key) {
+    return ((client *)key)->id;
+}
+
+int clientKeyCompare(dict *d, const void *key1, const void *key2) {
+    uint64_t l1,l2;
+    UNUSED(d);
+
+    l1 = ((client *)key1)->id;
+    l2 = ((client *)key2)->id;
+    return l1 == l2;
+}
+
+/* Set dictionary type. Keys are client, values are not used. */
+dictType clientDictType = {
```

Comment:
i also wonder what are our risks about this table going into rehashing now and then.
i didn't have time to think this through yet, maybe you did, so you can elaborate.

Reply:
I think it may affect the performance of `publish` command, because we need to iterate more buckets when rehashing. I will do some experiments to confirm it.

Reply:
I did some experiments about `publish` command and the results show that this PR won't affect the performance of `publish`. In my experiments, every client subscribes 100 channel. Then I publish messages(12 bytes) in random channel and record rps by using commands like `./src/redis-benchmark -c 50 -n 1000000 -r 100 publish __rand_int__  __rand_int__`.   
The results are as follow. The header is the number of the clients. 
||10|100|500|1000|5000|
|--|--|--|--|--|--|
|unstable(9d0158bf89265da)|94598.43|29728.28|6971.88|3556.31|441.87|
|optimize-pubsub|93571.62|29889.11|7042.58|3689.76|468.14 |

I also use the _perf_ and find that `addReplyBulk` takes more time than `listNext` in unstable branch or `dictNext` in this PR.